### PR TITLE
Realtime Eventing Sync

### PR DESCRIPTION
**Pull request by KavishShah09**

### CODE REVIEW

File: gateway/modules/eventing/eventing.go [hunk 1]
```diff
@@ -106,8 +106,5 @@ func (m *Module) SetConfig(project string, eventing *config.Eventing) error {
 		m.config.Rules = map[string]config.EventingRule{}
 	}
 
-	// Reset the internal rules
```
Comment (YourTechBud) [inline gateway/modules/eventing/eventing.go]: Set the internal rules if it was nil

File: gateway/modules/eventing/operations.go [hunk 2]
```diff
@@ -81,20 +82,10 @@ func (m *Module) AddInternalRules(eventingRules []config.EventingRule) {
...
-		isPresent := false
-		for _, storedRule := range m.config.InternalRules {
-
-			// Add the rule for the only if it doesn't already exist
-			if isRulesMatching(&storedRule, &incomingRule) {
-				isPresent = true
-				break
-			}
-		}
+	m.config.InternalRules = map[string]config.EventingRule{}
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations.go]: You can't do this. This will reset all the rules

Comment (YourTechBud) [reply gateway/modules/eventing/operations.go]: Change the name of the function to `SetRealtimeTriggers`

File: gateway/modules/eventing/operations.go [hunk 3]
```diff
@@ -81,20 +82,10 @@ func (m *Module) AddInternalRules(eventingRules []config.EventingRule) {
...
 
-		if !isPresent {
-			key := ksuid.New().String()
-			m.config.InternalRules[key] = incomingRule
-		}
+	for _, incomingRule := range eventingRules {
+		key := strings.Join([]string{"realtime", ksuid.New().String()}, "/")
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations.go]:
```suggestion
		key := strings.Join([]string{"realtime", ksuid.New().String()}, "-")
```

**Comment by YourTechBud** [comment]
Fix the linting errors

File: gateway/modules/eventing/eventing.go [hunk 4]
```diff
@@ -107,7 +107,9 @@ func (m *Module) SetConfig(project string, eventing *config.Eventing) error {
 	}
 
 	// Reset the internal rules
-	m.config.InternalRules = map[string]config.EventingRule{}
+	if m.config.Rules == nil {
```
Comment (YourTechBud) [inline gateway/modules/eventing/eventing.go]:
```suggestion
	if m.config.InternalRules == nil {
```

File: gateway/modules/eventing/operations.go [hunk 5]
```diff
@@ -81,20 +82,19 @@ func (m *Module) AddInternalRules(eventingRules []config.EventingRule) {
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
-	for _, incomingRule := range eventingRules {
-		isPresent := false
-		for _, storedRule := range m.config.InternalRules {
+	m.SetRealtimeTriggers(eventingRules)
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations.go]: I meant, rename `AddInternalRules` to `SetRealtimeTriggers`

File: gateway/modules/eventing/operations.go [hunk 6]
```diff
@@ -81,20 +82,19 @@ func (m *Module) AddInternalRules(eventingRules []config.EventingRule) {
...
+}
 
-			// Add the rule for the only if it doesn't already exist
-			if isRulesMatching(&storedRule, &incomingRule) {
-				isPresent = true
-				break
-			}
+// SetRealtimeTriggers sets the internal rules for realtime module
+func (m *Module) SetRealtimeTriggers(eventingRules []config.EventingRule) {
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations.go]: You'll need to take a lock here

File: gateway/modules/eventing/operations.go [hunk 7]
```diff
@@ -81,20 +82,19 @@ func (m *Module) AddInternalRules(eventingRules []config.EventingRule) {
...
+	for _, storedRule := range m.config.InternalRules {
+		if strings.HasPrefix(storedRule.Name, "realtime") {
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations.go]: use the key here instead of `storedRule.Name`

File: gateway/modules/eventing/helpers.go [hunk 8]
```diff
@@ -192,22 +192,22 @@ func (m *Module) getMatchingRules(name string, options map[string]string) []conf
 	return rules
 }
 
-func isRulesMatching(rule1 *config.EventingRule, rule2 *config.EventingRule) bool {
+// func isRulesMatching(rule1 *config.EventingRule, rule2 *config.EventingRule) bool {
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go]: Why did you comment this? If it isn't being used anymore then delete it

File: gateway/modules/eventing/operations_test.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package eventing
+
+import (
+	"testing"
+
+	"github.com/spaceuptech/space-cloud/gateway/config"
+)
+
+func TestModule_SetRealtimeTriggers(t *testing.T) {
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: Include the following test cases
- add a few eventing rules when no internal rules exist
- add a few eventing rules when a few non realtime eventing rules already exist
- add a few eventing rules when a few realtime eventing rule already exist
- add a few eventing rules when both realtime and non realtime rules already exist

In each test case, check if the final internal rules is set as expected

Comment (KavishShah09) [reply gateway/modules/eventing/operations_test.go]: How do i check the final result? My function doesn't return anything that i can assign as "got" to compare with want

Comment (YourTechBud) [reply gateway/modules/eventing/operations_test.go]: That's ok. You can check the struct directly

File: gateway/modules/eventing/operations_test.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		{
+			name: "no rules with prefix 'realtime'",
+			m:    &Module{config: &config.Eventing{InternalRules: map[string]config.EventingRule{"notrealtime": config.EventingRule{}}}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}}},
+			want: map[string]config.EventingRule{"notrealtime": config.EventingRule{}, "realtime-db-col-type": config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}},
+		},
+		{
+			name: "rules with prefix 'realtime'",
+			m:    &Module{config: &config.Eventing{InternalRules: map[string]config.EventingRule{"realtime-abc": config.EventingRule{}}}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{}}},
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: put the db and col here

File: gateway/modules/eventing/operations_test.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+			want: map[string]config.EventingRule{"realtime---": config.EventingRule{}},
+		},
+		{
+			name: "add eventing rules when no internal rules exist",
+			m:    &Module{config: &config.Eventing{InternalRules: make(map[string]config.EventingRule)}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, config.EventingRule{}}},
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: dont leave empty rule

File: gateway/modules/eventing/operations_test.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+			want: map[string]config.EventingRule{"realtime-db-col-type": config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, "realtime---": config.EventingRule{}},
+		},
+		{
+			name: "add eventing rules when no realtime internal rules exist",
+			m:    &Module{config: &config.Eventing{InternalRules: map[string]config.EventingRule{"notrealtime": config.EventingRule{}}}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, config.EventingRule{}}},
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: no empty rules please

File: gateway/modules/eventing/operations_test.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+			want: map[string]config.EventingRule{"notrealtime": config.EventingRule{}, "realtime-db-col-type": config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, "realtime---": config.EventingRule{}},
+		},
+		{
+			name: "add eventing rules when realtime internal rules exist",
+			m:    &Module{config: &config.Eventing{InternalRules: map[string]config.EventingRule{"realtime-abc": config.EventingRule{}, "realtime-def": config.EventingRule{}}}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, config.EventingRule{}}},
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: no empty rules please

File: gateway/modules/eventing/operations_test.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+			want: map[string]config.EventingRule{"realtime-db-col-type": config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, "realtime---": config.EventingRule{}},
+		},
+		{
+			name: "add eventing rules when realtime and non-realtime internal rules exist",
+			m:    &Module{config: &config.Eventing{InternalRules: map[string]config.EventingRule{"realtime-abc": config.EventingRule{}, "nonrealtime-def": config.EventingRule{}}}},
+			args: args{eventingRules: []config.EventingRule{config.EventingRule{Type: "type", Options: map[string]string{"db": "db", "col": "col"}}, config.EventingRule{}}},
```
Comment (YourTechBud) [inline gateway/modules/eventing/operations_test.go]: no empty rules please

**Pull request merged**